src/docfx/build/legacy/files/LegacyOutput.cs

```diff
@@ -2,33 +2,40 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Collections.Generic;
-using System.IO;
 using System.Threading.Tasks;
 
 namespace Microsoft.Docs.Build
 {
     internal static class LegacyOutput
     {
-        public static void Convert(Docset docset, Context context, List<(LegacyManifestItem manifestItem, Document document)> files, TableOfContentsMap tocMap)
+        public static async Task Convert(Docset docset, Context context, List<(LegacyManifestItem manifestItem, Document document)> files, TableOfContentsMap tocMap)
         {
-            Parallel.ForEach(files, file =>
+            using (Progress.Start("Convert Legacy Files"))
             {
-                var document = file.document;
-                var manifestItem = file.manifestItem;
-                switch (document.ContentType)
-                {
-                    case ContentType.TableOfContents:
-                        LegacyTableOfContents.Convert(docset, context, document, manifestItem.Output);
-                        break;
-                    case ContentType.Markdown:
-                    case ContentType.Redirection:
-                        LegacyMarkdown.Convert(docset, context, document, manifestItem.Output, tocMap);
-                        break;
-                    case ContentType.Resource:
-                        LegacyResource.Convert(docset, context, document, manifestItem.Output);
-                        break;
-                }
-            });
+                await ParallelUtility.ForEach(
```

Comment:
>ForEach [](start = 38, length = 7)

I am a bit confused: there is no async operation in all these code paths, how can we achieve any perf benefit by changing to `await`?

Reply:
you are right, change to parallel foreach

---
In reply to: [208220204](https://github.com/dotnet/docfx/pull/3161#discussion_r208220204) [](ancestors = 208220204)